mediatek: add support for Cudy TR3000 v1

LuisMitaHL opened the pull request:
Commit message:
```
Hardware:
 - SoC: MediaTek MT7981B
 - CPU: 2x 1.3 GHz Cortex-A53
 - Flash: 128 MiB SPI NAND
 - RAM: 512 MiB
 - WLAN: 2.4 GHz, 5 GHz (MediaTek MT7976CN, 802.11ax)
 - Ethernet: 1x 10/100/1000/2500 Mbps RTL8221B WAN, 1x10/100/1000 Mbps MT7981 LAN
 - USB 3.0 port
 - Buttons: 1 Reset button, 1 VPN slider
 - LEDs: 1x Red, 1x White
 - Serial console: internal test points, 115200 8n1
 - Power: 5 VDC, 3 A

MAC addresses:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| WAN     | 80:af:ca:xx:xx:x1 | label+1   |
| LAN     | 80:af:ca:xx:xx:x0 | label     |
| WLAN 2g | 80:af:ca:xx:xx:x0 | label     |
| WLAN 5g | 82:af:ca:xx:xx:x0 |           |
+---------+-------------------+-----------+

Installation:
The installation must be done via TFTP by disassembling the router. On other occasions Cudy has distributed intermediate firmware to make installation easier, and so I recommend checking the Wiki for this device if there is a more convenient solution than the one below.

To install using TFTP:
1. Connect to UART.
2. With the router off, press the RESET button. While the router is turning on, the button should continue to be pressed for at least 5 seconds.
3. A u-boot shell will automatically open.
4. Connect to LAN and set your IP to 192.168.1.88/24. Configure a TFTP server and an OpenWrt initramfs-kernel.bin firmware file.
5. Run these steps in u-boot using the name of your file.

setenv bootfile initramfs-kernel.bin
tftpboot
bootm

6. If you can reach LuCI or SSH now, just use the sysupgrade image with the 'Keep settings' option turned off.

```

Tested:
- Linux boot
- UART
- Ethernet WAN (including correct MAC)
- Ethernet LAN (including correct MAC)
- Wi-Fi 2.4GHz, 5GHz (including correct MAC)
- USB Power & Data
- LEDs
- SPI Flash
- Failsafe
- Configurable slider

musashino205 commented on target/linux/mediatek/dts/mt7981b-cudy-tr3000-v1.dts:
```diff
@@ -0,0 +1,231 @@
+                        color = <LED_COLOR_ID_WHITE>;
+                        gpios = <&pio 10 GPIO_ACTIVE_LOW>;
+                };
+
+
```
Please remove these unnecessary newlines.

enmaskarado replied:
Done

musashino205 commented on target/linux/mediatek/dts/mt7981b-cudy-tr3000-v1.dts:
```diff
@@ -0,0 +1,231 @@
+                        color = <LED_COLOR_ID_RED>;
+                        gpios = <&pio 11 GPIO_ACTIVE_LOW>;
+                };
+                led@1 {
```
Please insert a newline between those nodes.

enmaskarado replied:
Done

musashino205 commented on target/linux/mediatek/dts/mt7981b-cudy-tr3000-v1.dts:
```diff
@@ -0,0 +1,231 @@
+#include "mt7981.dtsi"
+
+/ {
+        model = "Cudy TR3000 v1";
```
Please use tabs instead of spaces for indentation in dts.

enmaskarado replied:
Done. My original file had tabs, but they were converted to spaces by copying and pasting from the terminal.

musashino205 commented on target/linux/mediatek/dts/mt7981b-cudy-tr3000-v1.dts:
```diff
@@ -0,0 +1,231 @@
+                        groups = "spi0", "spi0_wp_hold";
+                };
+        };
+
```
Please remove this unnecessary newline.

enmaskarado replied:
Done

musashino205 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -472,6 +472,23 @@ define Device/cudy_re3000-v1
+  DEVICE_VARIANT := v1
+  DEVICE_DTS := mt7981b-cudy-tr3000-v1
+  DEVICE_DTS_DIR := ../dts
+  SUPPORTED_DEVICES += R47
```
Is this required?

enmaskarado replied:
I am not really sure. I got the R47 code from a DTS that Cudy has published in https://www.cudy.com/openwrt_software_download

Other Cudy routers in OpenWrt have the same variable too (and I used their DTS as an example for this)

musashino205 replied:
I see, it's for compatibility on the firmware validation when upgrading from an OpenWrt image provided by Cudy.
That image for TR3000 v1 is not provided now, but if there is a possibility that it will be released in the future, I'm fine to keep this `SUPPORTED_DEVICES`.

nicefile commented:
and one more thing.  To have working network in the recovery you should add entry to `/lib/preinit/05_set_preinit_iface` before "glinet,gl-mt3000" as it will enable eth1 port.

enmaskarado commented:
> and one more thing. To have working network in the recovery you should add entry to
> 
> ```
> /lib/preinit/05_set_preinit_iface
>    8         glinet,gl-mt3000)
>    9                 ip link set eth1 up
>   10                 ifname=eth1
> ```
> 
> before "glinet,gl-mt3000" as it will enable eth1 port.

Thank you! I noticed yesterday that the recovery wasn't working, and I was checking other GitHub issues today to see what the bug might be.

I tested it and failsafe is working fine now on the LAN port.

RaylynnKnight commented:
> Commit message:
> 
> ```
> Hardware:
>  - SoC: MediaTek MT7981B
>  - CPU: 2x 1.3 GHz Cortex-A53
>  - Flash: 128 MiB SPI NAND
>  - RAM: 512 MiB
>  - WLAN: 2.4 GHz, 5 GHz (MediaTek MT7976CN, 802.11ax)
>  - Ethernet: 1x 10/100/1000/2500 Mbps RTL8221B WAN, 1x10/100/1000 Mbps MT7981 LAN
>  - USB 3.0 port
>  - Buttons: 1 Reset button, 1 VPN slider
>  - LEDs: 1x Red, 1x White
>  - Serial console: internal 4-pin header, 115200 8n1
It might be preferable to change the above line to make clear that it is not a 4-pin header, but test points for the serial console!

enmaskarado commented:
> It might be preferable to change the above line to make clear that it is not a 4-pin header, but test points for the serial console!

Thank you. I didn't notice that copy and paste error.

enmaskarado commented:
I have modified how the slider button is recognized, and it now follows the behavior [indicated in the Wiki](https://openwrt.org/docs/guide-user/hardware/hardware.button).

Additionally I have changed `interrupts = <38 IRQ_TYPE_LEVEL_HIGH>;` to `interrupts = <38 IRQ_TYPE_LEVEL_LOW>;` following [Cudy's documentation](https://www.cudy.com/filedownload/781388).

LuisMitaHL committed:
mediatek: add support for Cudy TR3000 v1

Hardware:
 - SoC: MediaTek MT7981B
 - CPU: 2x 1.3 GHz Cortex-A53
 - Flash: 128 MiB SPI NAND
 - RAM: 512 MiB
 - WLAN: 2.4 GHz, 5 GHz (MediaTek MT7976CN, 802.11ax)
 - Ethernet: 1x 10/100/1000/2500 Mbps RTL8221B WAN, 1x10/100/1000 Mbps MT7981 LAN
 - USB 3.0 port
 - Buttons: 1 Reset button, 1 slider button
 - LEDs: 1x Red, 1x White
 - Serial console: internal test points, 115200 8n1
 - Power: 5 VDC, 3 A

MAC addresses:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| WAN     | 80:af:ca:xx:xx:x1 | label+1   |
| LAN     | 80:af:ca:xx:xx:x0 | label     |
| WLAN 2g | 80:af:ca:xx:xx:x0 | label     |
| WLAN 5g | 82:af:ca:xx:xx:x0 |           |
+---------+-------------------+-----------+

Installation:
The installation must be done via TFTP by disassembling the router. On other occasions Cudy has distributed intermediate firmware to make installation easier, and so I recommend checking the Wiki for this device if there is a more convenient solution than the one below.

To install using TFTP:
1. Connect to UART.
2. With the router off, press the RESET button. While the router is turning on, the button should continue to be pressed for at least 5 seconds.
3. A u-boot shell will automatically open.
4. Connect to LAN and set your IP to 192.168.1.88/24. Configure a TFTP server and an OpenWrt initramfs-kernel.bin firmware file.
5. Run these steps in u-boot using the name of your file.

setenv bootfile initramfs-kernel.bin
tftpboot
bootm

6. If you can reach LuCI or SSH now, just use the sysupgrade image with the 'Keep settings' option turned off.

Signed-off-by: Luis Mita <luis@luismita.com>

enmaskarado commented:
I upgraded to the latest version of main, with kernel 6.6.
Everything seems to work fine except the WAN port, which does not recognize when something is connected to it, or not.

This dmesg line catches my attention:
[   14.516421] mtk_soc_eth 15100000.ethernet eth0: PHY [mdio-bus:01] driver [Generic Clause 45 PHY] (irq=POLL)

RTL8221B-VB-CG used to appear as the chip model.

Full dmesg: https://pastebin.com/g3y8DaVs

I am not sure if this bug is from the PR or from the main project.

enmaskarado commented:
Looks like i'm hitting #15093

hauke commented on target/linux/mediatek/dts/mt7981b-cudy-tr3000-v1.dts:
```diff
@@ -0,0 +1,232 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_status: led@0 {
```
Please use `led_0`.

The `@` is used when there is a `reg` property and it should match that `reg` property. The LEDs do not have a `reg` property and then using a name without `@` is better. 

same for `led@1`

enmaskarado replied:
Ok, thanks

Samxieerbao commented:
你好，

256MB Flash版本的并未更换Flash,仍然可以用之前的固件，不需要打补丁。




Sam
Cudy  - Technical Support Engineer
wechat：Battista8888
teams:  ff5370f951d616d3
whatsapp:   +8618002593468
email:   ***@***.***
7th Floor, West Tower, Lepu Building, No. 66 Xingke Road, Xili Community, Xili Street, Nanshan District,Shenzhen, China 518055
 
From: cheungxiongwei
Date: 2025-12-02 00:26
To: openwrt/openwrt
CC: Samxieerbao; Comment
Subject: Re: [openwrt/openwrt] mediatek: add support for Cudy TR3000 v1 (PR #15214)
cheungxiongwei left a comment (openwrt/openwrt#15214)
您好，开发者， 我是 Cudy 的技术支持工程师 Sam。我们最近更改了 Flash（ROM）ESMT F50L1G41LC 的型号。 对于 ubootmod 布局版本，我们需要在 uboot 中添加一个新的补丁。补丁已附上，请将其粘贴到此处。 package/boot/uboot-mediatek/patches/100-30-mtd-spinand-esmt-Add-support-for-F50L1G41LC.patch 100-30-mtd-spinand-esmt-Add-support-for-F50L1G41LC.patch
你好，Sam，
我最近购买了 Cudy TR3000 256MB FLASH。
SN：TR3000 2544 xxxxx
我使用合并了补丁的 openwrt-23.05.5-mediatek-filogic-cudy_tr3000-v1-squashfs-sysupgrade.bin 2025年12月1号的快照版本，在贵公司发布的过渡固件上，进行的升级，网络没问题，但是无法打开 http://192.168.1.1/ 主页。可使用 ssh 进入路由器。
***@***.***:/proc# cat version
Linux version 6.12.59 ***@***.***) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 14.3.0 r32183-0e7a7e62fb) 14.3.0, GNU ld (GNU Binutils) 2.44) #0 SMP Sun Nov 30 23:51:31 2025
***@***.***:/proc# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00100000 00020000 "BL2"
mtd1: 00080000 00020000 "u-boot-env"
mtd2: 00200000 00020000 "Factory"
mtd3: 00040000 00020000 "bdinfo"
mtd4: 00200000 00020000 "FIP"
mtd5: 07240000 00020000 "ubi"

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***>
